### ISSUE TITLE
DataParallel is used by auto_model with single GPU

### ISSUE THREAD

**Issue by H4dr1en** [pytorch/ignite#2447]
## 🐛 Bug description

I am not sure whether it is a bug or a feature:

The DataParallel is being applied/patched by `idist.auto_model` in the context of a single gpu (backend=None, nproc_per_node=1). What is the reason behind this choice? Does it bring any speed improvements?

The only way to prevent it is to set `os.environ["CUDA_VISIBLE_DEVICES"] = "0"` for single-gpu contexts.

## Environment

- PyTorch Version (e.g., 1.4): 1.7.1
- Ignite Version (e.g., 0.3.0): 0.4.8
- OS (e.g., Linux): Linux
- How you installed Ignite (`conda`, `pip`, source): pip
- Python version: 3.8

**Comment by vfdev-5** [comment]
@H4dr1en I haven't yet checked the code but I agree that this seems useless to wrap the model by `DataParallel` in case of 1 device.

**Comment by H4dr1en** [comment]
> @H4dr1en I haven't yet checked the code but I agree that this seems useless to wrap the model by DataParallel in case of 1 device. [...] I think this is expected behaviour if you have N GPUs, start the script without any restriction on how many GPUs to use and use idist.auto_model.

DataParallel seems to drastically slow down the model training. I tried with the cifar10 example from the pytorch-ignite/examples repository on [this fork](https://github.com/H4dr1en/examples/commit/4e7a1bb91d3e44bc200533157796735aac5445ae), I ran this script on a g4dn.12xlarge (x4 T4 GPUs AWS instance).

The default training time reported by ignite is 
```
CIFAR10-Training INFO: Engine run complete. Time taken: 00:29:53
```
If I set `os.environ["CUDA_VISIBLE_DEVICES"] = "0"`(line 430), the training time becomes
```
CIFAR10-Training INFO: Engine run complete. Time taken: 00:05:34
```
This is the only change I made. If this is truly coming from the DataParallel being not efficient for a single GPU, I see no reason why it should be automatically set. I'd expect the `idist.auto_model` to give me the best/fastest configuration, in this case, no `DataParallel`. Am I missing something?

**Comment by vfdev-5** [comment]
Yeah, I was also confused about how you are using `nproc_per_node` in your code. By the way, I still do not quite understand what happens exactly if `nproc_per_node > 1`, what happens with the subprocess entering into `__main__`, `config` has `nproc_per_node=4` by default, so how it works that it does respawn inf number of processes ?

>  I expect to have the training on one GPU, but it is actually training on all the GPUs available with DDP, which make it slower. Is it correct?

I'd expect training on 4 GPUs in DDP mode (not DP) is faster than 1 GPU. In your case, I suppose the slowdown with `nproc_per_node=1` and using 4 GPUs in DP mode is related to a small batch size.

**Comment by H4dr1en** [comment]
> By the way, I still do not quite understand what happens exactly if nproc_per_node > 1, what happens with the subprocess entering into __main__, config has nproc_per_node=4 by default, so how it works that it does respawn inf number of processes ?

The subprocesses have args.local_rank defined, so they just start the training

> I'd expect training on 4 GPUs in DDP mode (not DP) is faster than 1 GPU. In your case, I suppose the slowdown with nproc_per_node=1 and using 4 GPUs in DP mode is related to a small batch size.

Yes, on the same page. What I want to raise in this issue is the fact that `idist.auto_model`, when being executed with an environment where `nproc_per_node=1`, regardless of the number of GPUs available, should not use DP, and only train using a single GPU. - does it make sense?

**Comment by vfdev-5** [comment]
> What I want to raise in this issue is the fact that idist.auto_model, when being executed with an environment where nproc_per_node=1, regardless of the number of GPUs available, should not use DP, and only train using a single GPU - does it make sense?

Yes, it makes perfectly sense, but how ignite can understand that you have `nproc_per_node=1` ?

**Comment by H4dr1en** [comment]
> Yes, it makes perfectly sense, but how ignite can know that you have nproc_per_node=1 ?

Can we maybe check the world_size?

> In addition, there can be (old) cases when we would like to use DP : one process and use multiple GPUs.

Is it justified to keep this use case, now that DDP is out and is faster than DP? I don't fully understand the different use cases, so I might be wrong, in which case I understand that we should not change this behaviour

**Comment by vfdev-5** [comment]
> Can we maybe check the world_size?

yes, we are using world_size to setup DDP. If world_size is defined and >1 then there is a distributed processing group and there is no point to use DP.

Here is the code: https://github.com/pytorch/ignite/blob/6d83dd72bb0bb7e655cd284789f367b46ab36a9e/ignite/distributed/auto.py#L201-L230
